WMF Framework/Significant Events Endpoint/SignificantEventsModels.swift

```diff
@@ -0,0 +1,616 @@
+
+import Foundation
+enum SignificantEventsDecodeError: Error {
+    case unableToParseIntoTypedEvents
+}
+
+public struct SignificantEvents: Decodable {
+    public let nextRvStartId: UInt?
+    public let sha: String?
+    private let untypedEvents: [UntypedEvent]
+    public let typedEvents: [TypedEvent]
+    public let summary: Summary
+    
+    enum CodingKeys: String, CodingKey {
+        case nextRvStartId
+        case sha
+        case untypedEvents = "timeline"
+        case typedEvents
+        case summary
+    }
+    
+    public struct Summary: Decodable {
+        public let earliestTimestampString: String
+        public let numChanges: UInt
+        public let numUsers: UInt
+        
+        enum CodingKeys: String, CodingKey {
+            case earliestTimestampString = "earliestTimestamp"
+            case numChanges
+            case numUsers
+        }
+    }
+    
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        nextRvStartId = try? container.decode(UInt.self, forKey: .nextRvStartId)
+        sha = try? container.decode(String.self, forKey: .sha)
+        summary = try container.decode(Summary.self, forKey: .summary)
+        untypedEvents = try container.decode([UntypedEvent].self, forKey: .untypedEvents)
+        
+        var typedEvents: [TypedEvent] = []
+        
+        for untypedEvent in untypedEvents {
+            switch untypedEvent.outputType {
+            case .small:
+                if let event = Event.Small(untypedEvent: untypedEvent) {
+                    typedEvents.append(.small(event))
+                }
+            case .large:
+                if let event = Event.Large(untypedEvent: untypedEvent) {
+                    typedEvents.append(.large(event))
+                }
+            case .vandalismRevert:
+                if let event = Event.VandalismRevert(untypedEvent: untypedEvent) {
+                    typedEvents.append(.vandalismRevert(event))
+                }
+            case .newTalkPageTopic:
+                if let event = Event.NewTalkPageTopic(untypedEvent: untypedEvent) {
+                    typedEvents.append(.newTalkPageTopic(event))
+                }
+            }
+        }
+
+        guard typedEvents.count == untypedEvents.count else {
```

Comment:
Yeah I agree, it's too rigid to fail if one event isn't typed. I changed this to error out against a case where there are no typed events at this point whatsoever, which is probably pretty rare but I wanted to avoid any empty events state that might happen at the UI level.

Reply:
Cool, that sounds like a good failure case here.